### PR TITLE
msgs/<username> api endo

### PR DESCRIPTION
**Pull request by KnittedSox**

### CODE REVIEW

File: minitwit/src/main/java/WebApplication.java [hunk 1]
```diff
@@ -665,16 +667,67 @@ else if (!BCrypt.checkpw(enteredPW, rs.getString("pw_hash"))) {
...
+            var time = System.currentTimeMillis() / 1000L;
+
+            var query = connection.prepareStatement("INSERT INTO message (author_id, text, pub_date, flagged) VALUES (?, ?, ?, 0)");
+            query.setInt(1, userID);
+            query.setString(2, text);
+            query.setLong(3, time);
+
+            query.execute();
+            connection.close();
+            return "204";
```
Comment (pizzaluc) [inline minitwit/src/main/java/WebApplication.java:726]: call this before returning
response.status(204);

File: minitwit/src/main/java/WebApplication.java [hunk 2]
```diff
@@ -665,16 +667,67 @@ else if (!BCrypt.checkpw(enteredPW, rs.getString("pw_hash"))) {
...
+
+        SQLite db = new SQLite();
+        var connection = db.getConnection();
+        var userID = getUserID(db, request.params(":username"));
+
+        int noMsgs =  Integer.parseInt(request.queryParamOrDefault("no", "100"));
+
+        if (request.requestMethod().equals("GET")){
+            if (userID == 0) {
+                halt(404, "404: Could not find user!"); //How do we treat 404?
```
Comment (pizzaluc) [inline minitwit/src/main/java/WebApplication.java:694]: I think we can just return
response.status(404);

Or is there another status that reflects better that the user does not exist

Comment (KnittedSox) [reply minitwit/src/main/java/WebApplication.java:694]: Hmm maybe. In the API specification it uses flask's abort. Not entirely sure that halt functions the same though, but it seemed the most relevant to me .

File: minitwit/src/main/java/WebApplication.java [hunk 3]
```diff
@@ -665,16 +667,67 @@ else if (!BCrypt.checkpw(enteredPW, rs.getString("pw_hash"))) {
...
+    //  --------------------------------------
+    //              Simulator API
+    //  ---------------------------------------
+
+    public static Route serveSimMsgsUsr = (Request request, Response response) -> {
+
+        updateLatest(request);
+
+        SQLite db = new SQLite();
+        var connection = db.getConnection();
```
[the remark this answers is not in the document]
Comment (KnittedSox) [reply minitwit/src/main/java/WebApplication.java:687]: Good catch. Fixing it now

File: minitwit/src/main/java/WebApplication.java [hunk 4]
```diff
@@ -102,23 +103,24 @@ public static void main(String[] args) {
...
         get(URLS.UNFOLLOW, WebApplication.serveUnfollowPage);
-        get(URLS.USER, WebApplication.serveUserTimelinePage);
-        get(URLS.USER_TIMELINE, WebApplication.serveUserByUsernameTimelinePage);
         post(URLS.LOGIN, WebApplication.serveLoginPage);
 
         // Sim API
         path("/api", () -> {
             // All endpoints in this path must be authenticated
-            before("/*", protectEndpoint);
+            //before("/*", protectEndpoint);
```
[the remark this answers is not in the document]
Comment (KnittedSox) [reply minitwit/src/main/java/WebApplication.java:118]: Nope, thought I changed it back. Fixing now

File: minitwit/src/main/java/WebApplication.java [hunk 5]
```diff
@@ -665,16 +668,69 @@ else if (!BCrypt.checkpw(enteredPW, rs.getString("pw_hash"))) {
...
+                    filteredMessage.put("user", messages.getString("username"));
+                    filteredMessages.add(filteredMessage);
+                }
+                connection.close();
+                return filteredMessages.stream().toList();
+            }
+        }
+        // Might pose a problem when we POST a message for a user that does not exist (lots of author_id=0 messages) not sure
+        else if (request.requestMethod().equals("POST")) {
+            var text = request.queryParams("content");
```
Comment (Herover) [inline minitwit/src/main/java/WebApplication.java:719]: The content should come from the body and not a query parameter, eg. `curl -X POST -H "Content-Type: application/json" -d '{"content": "Hello"}' --header "Authorization: Basic c2ltdWxhdG9yOnN1cGVyX3NhZmUh" http://localhost:8080/api/msgs/asdf?latest=123`

File: minitwit/src/main/java/WebApplication.java [hunk 6]
```diff
@@ -665,16 +668,69 @@ else if (!BCrypt.checkpw(enteredPW, rs.getString("pw_hash"))) {
...
+
+        SQLite db = new SQLite();
+        var connection = db.getConnection();
+        var userID = getUserID(db, request.params(":username"));
+
+        int noMsgs =  Integer.parseInt(request.queryParamOrDefault("no", "100"));
+
+        if (request.requestMethod().equals("GET")){
+            if (userID == 0) {
+                response.status(404);
```
Comment (Herover) [inline minitwit/src/main/java/WebApplication.java:695]: connection is not closed here.
Also, if we POST with a invalid user this one doesn't catch that, but that bug also exists in the spec api :D

**Review by Herover — CHANGES_REQUESTED** [review]

**Review by smilladion — APPROVED** [review]
Works perfectly for me now! I just added a commit which returns empty string. adds status codes and closes db in certain places.

**Review by Herover — APPROVED** [review]

**Pull request merged**